### ISSUE TITLE
to_string

### ISSUE THREAD

**Issue by yoni206** [pysathq/pysat#113]
It would be great to have a way of pretty-printing formulas, by providing a dictionary between the variables (numerals) and application-specific name. This is easy to do for ordinary CNF using dictionaries, but I am not sure how to do this for pseudo-Boolean constraints. This would be nice to have for debugging applications that use pysat.

**Comment by alexeyignatiev** [comment]
Good point, @yoni206. But what do you mean by "application-specific name"? The standard (and most straightforward) way to implement formula's `__str__()` is to make it return its DIMACS representation.

**Comment by alexeyignatiev** [comment]
Have a look at the functionality of [IDPool](https://pysathq.github.io/docs/html/api/formula.html#pysat.formula.IDPool). It is meant to represent this kind of dictionary.

**Comment by RexYuan** [comment]
@yoni206 do you mean that, for ordinary CNF, given clauses
```
[
  [1, -2, 3],
  [2, -3]
]
```
and dictionary
```
{
  1: "a",
  2: "b",
  3: "c"
}
```
you want to print out something like this?
```
(a | ~b | c) & (b | ~c)
```

**Comment by alexeyignatiev** [comment]
I guess this string may help a non-SAT researcher understand the meaning of the formula but a SAT practitioner would perhaps prefer a formula in DIMACS. It has been a standard for many years and everybody speaks this "language". What I meant above was that the mapping from integer variable identifiers to the corresponding objects can be provided by means of the already existing `IDPool`. I presume it can be added into comments line sitting in a CNF object.

**Comment by yoni206** [comment]
@RexYuan yes, but also for cardinality constraints.

**Comment by RexYuan** [comment]
@yoni206 How about using two dictionaries like this?
```
card_clauses = [
  [[1, -2, 3], 2],
  [[2, -3], 1]
]
card_names = {
  2: "n",
  1: "m"
}
var_names = {
  1: "a",
  2: "b",
  3: "c"
}
```
then you have
```
a + ~b + c <= n
b + ~c <= m
```
Of course, this would run into a problem if there are different names for a cardinality, but I think then they are already indistinguishable on the clause level.

**Comment by yoni206** [comment]
yeah that could work
